src/internal/partial_solution.rs

```diff
@@ -1,103 +1,201 @@
 use crate::package::Package;
 use crate::range::Range;
 use crate::term::Term;
 use crate::type_aliases::{Map, SelectedDependencies};
 use crate::version::Version;
 
-#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq)]
-pub struct DecisionLevel(u32);
+use super::small_vec::SmallVec;
 
-impl std::ops::Add<DecisionLevel> for DecisionLevel {
-    type Output = DecisionLevel;
+#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq)]
+pub struct DecisionLevel(pub u32);
 
-    fn add(self, other: DecisionLevel) -> DecisionLevel {
-        DecisionLevel(self.0 + other.0)
+impl DecisionLevel {
+    pub fn increment(self) -> Self {
+        Self(self.0 + 1)
     }
 }
 
-impl std::ops::SubAssign<DecisionLevel> for DecisionLevel {
-    fn sub_assign(&mut self, other: DecisionLevel) {
-        self.0 -= other.0
-    }
+/// The partial solution contains all package assignments,
+/// organized by package and historically ordered.
+#[derive(Clone, Debug)]
+pub struct PartialSolution<P: Package, V: Version> {
+    next_global_index: u32,
+    current_decision_level: DecisionLevel,
+    package_assignments: Map<P, PackageAssignments<P, V>>,
 }
 
-#[derive(Clone)]
-pub struct DatedAssignment<P: Package, V: Version> {
-    decision_level: DecisionLevel,
-    assignment: Assignment<P, V>,
+/// Package assignments contain the potential decision and derivations
+/// that have already been made for a given package,
+/// as well as the intersection of terms by all of these.
+#[derive(Clone, Debug)]
+struct PackageAssignments<P: Package, V: Version> {
+    smallest_decision_level: DecisionLevel,
+    highest_decision_level: DecisionLevel,
+    dated_derivations: SmallVec<DatedDerivation<P, V>>,
+    assignments_intersection: AssignmentsIntersection<V>,
 }
 
-/// The partial solution is the current state
-/// of the solution being built by the algorithm.
-/// It is composed of a succession of assignments,
-/// defined as either decisions or derivations.
-#[derive(Clone)]
-pub struct PartialSolution<P: Package, V: Version> {
+#[derive(Clone, Debug)]
+pub struct DatedDerivation<P: Package, V: Version> {
+    global_index: u32,
     decision_level: DecisionLevel,
-    /// Each assignment is stored with its decision level in the history.
-    /// The order in which assignments where added in the vec is kept,
-    /// so the oldest assignments are at the beginning of the vec.
-    history: Vec<DatedAssignment<P, V>>,
-    memory: Memory<P, V>,
+    cause: IncompId<P, V>,
+}
+
+#[derive(Clone, Debug)]
+enum AssignmentsIntersection<V: Version> {
+    Decision((u32, V, Term<V>)),
+    Derivations(Term<V>),
+}
+
+/// An assignment is either a decision: a chosen version for a package,
+/// or a derivation : a term specifying compatible versions for a package.
+/// We also record the incompatibility at the origin of a derivation, called its cause.
+/// TODO: Remove. I only added that to not change the code in core.rs (except reference &).
```

Comment:
That comment was here because I wanted to reduce the size of the git diff for now but not forget to improve the code by removing that type later, because actually, we don't really need that type. With the history/memory unification, it's only useful in one place now, as the return type of the `find_satisfier_and_previous_satisfier_level` function. And by keeping the same return type, we don't have to modify the code using it in the `core` module.

But in fact, I think it would be cleaner to remove that `Assignment` type and change the return type of that function to something like `(&P, SatisfierSearch)` where that type would be defined with something like
```rust
enum SatisfierSearch<P, V> {
    DifferentDecisionLevels { previous_satisfier_level: DecisionLevel },
    SameDecisionLevels { satisfier_cause: IncompId<P, V> },
}
```

Reply:
looks to me like we need it to get passed to `prior_cause`, but we can leave it for latter.